### PR TITLE
feat: add cloud logs support

### PR DESCRIPTION
**Pull request by iamar7**
### Description

Resolves: https://github.com/terraform-ibm-modules/stack-ibm-core-security-services/issues/115

### Release required?

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [x] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### For mergers

- Use a conventional commit message to set the release level. Follow the [guidelines](https://terraform-ibm-modules.github.io/documentation/#/merging.md).
- Include information that users need to know about the PR in the commit message. The commit message becomes part of the GitHub release notes.
- Use the **Squash and merge** option.

### CODE REVIEW

File: stack_definition.json [hunk 1]
```diff
@@ -173,10 +180,18 @@
         {
           "name": "enable_platform_metrics",
           "value": "ref:../../inputs/enable_platform_logs_metrics"
+        },
+        {
+          "name": "existing_en_instance_crn",
+          "value": "ref:../../inputs/existing_en_instance_crn"
```
Comment (ocofaigh) [inline stack_definition.json:186]: thats not right - the ref needs to come from the Event Notifications DA output

**Review by ocofaigh — CHANGES_REQUESTED** [review]
- You need to rename `enable_platform_logs_metrics` to `enable_platform_metrics` since we will no longer provision Log Analysis. To enable platform logs with cloud logs, users will not use the new `logs_routing_tenant_regions` input.
- There is now a circular dependency you need to resolve - removed `existing_monitoring_crn` from Event Notifications input (line 239). The dependency is now the other way around - Observability depends on EN.

**Comment by iamar7** [comment]
Getting this in the pipeline.
> Error: ServiceNotAuthorized: The specified COS Service Instance does not have sufficient permissions to access the resource associated with the KMS key CRN.
>  2024/09/27 11:26:44 Terraform apply | 	status code: 401, request id: 2741b9da-xxxd-4xx2-bxxa-b5xxxxxXXxxxxdbdf, host id: 
>  2024/09/27 11:26:44 Terraform apply | 
>  2024/09/27 11:26:44 Terraform apply |   with module.cos[0].ibm_cos_bucket.cos_bucket[0],
>  2024/09/27 11:26:44 Terraform apply |   on .terraform/modules/cos/main.tf line 133, in resource "ibm_cos_bucket" "cos_bucket":
>  2024/09/27 11:26:44 Terraform apply |  133: resource "ibm_cos_bucket" "cos_bucket" {

**Review by ocofaigh — CHANGES_REQUESTED** [review]
You can keep `enable_platform_logs_metrics` instead of splitting into 2 separate ones

**Comment by iamar7** [comment]
Getting this in the pipeline run -

Screenshot 2024-10-03 at 6 16 42 AM

**Comment by ocofaigh** [comment]
@iamar7 Can we update all of the version locators to match what we have in the RAG stack: https://github.com/terraform-ibm-modules/stack-retrieval-augmented-generation/blob/main/solutions/basic/stack_definition.json

At least we know that exact combination works. Keep on eye on some of the member inputs and refs, as they may need updates too.

**Comment by iamar7** [comment]
@ocofaigh The pipeline is failing is due to the changes in order of `observability` and `event notification`. Earlier, the `cos_kms_auth_policy` was being created in the observability-da and skipped in the event-notification DA. So now since the event notification DA is being deployed first so need to create the `cos_kms_auth_policy` in the `event-notification DA` and skip it in the observability-da.

**Comment by ocofaigh** [comment]
I think we will want to do a manual upgrade test against this branch before we proceed - but lets get pipeline passing here fist and ill get prepped for a manual upgrade

**Comment by iamar7** [comment]
`TestProjectsFullTest` passed in the above run.

**Comment by ocofaigh** [comment]
As discussed internally, there is a breaking change here due to the dependency tree order changing and the requirement to change which DA is creating the COS to KMS auth policy. We are discussing options currently.

**Review by ocofaigh — CHANGES_REQUESTED** [review]
@iamar7 While we are making a decision on how to handle the breaking change, can you please update the diagram to replace Log Analysis with Cloud Logs, the cloud logs bucket, and the AT route to cloud logs

**Comment by iamar7** [comment]
> @iamar7 While we are making a decision on how to handle the breaking change, can you please update the diagram to replace Log Analysis with Cloud Logs, the cloud logs bucket, and the AT route to cloud logs

Sure @ocofaigh

File: stack_definition.json [hunk 2]
```diff
@@ -219,14 +227,10 @@
         {
           "name": "prefix",
           "value": "ref:../../inputs/prefix"
-        },
-        {
-          "name": "existing_monitoring_crn",
-          "value": "ref:../../members/2 - Observability/outputs/cloud_monitoring_crn"
```
Comment (ocofaigh) [inline stack_definition.json:225]: FYI, I found another projects [bug](https://github.ibm.com/epx/projects/issues/3646) where projects is not removing this value from the stack on upgrade. So for now, I think we need to workaround it by explicitly passing `__NULL__` here. I'm going to push a commit and try it out.

**Comment by ocofaigh** [comment]
There is another issue we are seeing now when upgrading Observability to remove logdna:
```
 2024/10/08 07:47:24 Terraform plan | Error: Cannot read the remote archive resource
 2024/10/08 07:47:24 Terraform plan | 
 2024/10/08 07:47:24 Terraform plan |   with module.observability_instance.module.log_analysis[0].logdna_archive.archive_config[0],
 2024/10/08 07:47:24 Terraform plan |   on .terraform/modules/observability_instance/modules/log_analysis/main.tf line 42, in resource "logdna_archive" "archive_config":
 2024/10/08 07:47:24 Terraform plan |   42: resource "logdna_archive" "archive_config" {
 2024/10/08 07:47:24 Terraform plan | 
 2024/10/08 07:47:24 Terraform plan | expected either servicekey or iamtoken to be set
```

My guess is since we have identified the logdna resources (including the resource key used to initialise the logdna provider) for deletion, the logdna provider can no longer be initialised, and we get this error.
We need to find a way to handle this.

**Comment by ocofaigh** [comment]
Going to have to do the upgrade in a 2 step process to fully removed Log Analysis. In this PR we will disable the log archiving, but keep the LogDNA instance up, as well as deploy Cloud Logs. We will deploy Cloud Logs without Event Notifications integration for now, and in the next release, we will add it, since it will require manual steps to upgrade, so hence will make the next change a major version update.

**Comment by ocofaigh** [comment]
Some issue with key protect service:
```
 2024/10/11 10:05:51 Terraform apply | ---
 2024/10/11 10:05:51 Terraform apply | id: terraform-46882f4c
Error: 0/11 10:05:51 Terraform apply | summary: '[ERROR] Error while creating key ring : kp.Error:
 2024/10/11 10:05:51 Terraform apply | correlation_id=''7c640068-1d71-4343-859f-99bbe462ece7'',
 2024/10/11 10:05:51 Terraform apply |   msg=''This operation cannot be completed. Please try your request at a later time.
 2024/10/11 10:05:51 Terraform apply |   If the issue persists, please contact IBM KeyProtect.'''
 2024/10/11 10:05:51 Terraform apply | severity: error
 2024/10/11 10:05:51 Terraform apply | resource: ibm_kms_key_rings
 2024/10/11 10:05:51 Terraform apply | operation: create
 2024/10/11 10:05:51 Terraform apply | component:
 2024/10/11 10:05:51 Terraform apply |   name: github.com/IBM-Cloud/terraform-provider-ibm
 2024/10/11 10:05:51 Terraform apply |   version: 1.69.2
 2024/10/11 10:05:51 Terraform apply | ---
```

Retrying..

**Review by ocofaigh — APPROVED** [review]

**Pull request merged**